File: algorithms/algorithms-cplusplus/leetcode/Question_0349_Intersection_of_Two_Arrays.cc
#include <vector>
#include <set>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
    vector<int> res;
    sort(nums1.begin(), nums1.end());
    sort(nums2.begin(), nums2.end());
    int i = 0, j = 0;
    while (i < nums1.size() && j < nums2.size()) {
      if (nums1[i] == nums2[j]) {
        res.push_back(nums1[i]);
        while (i < nums1.size() - 1 && nums1[i] == nums1[i + 1]) {
          i++;
        }
        while (j < nums2.size() - 1 && nums2[j] == nums2[j + 1]) {
          j++;
        }
        i++; j++;
      } else if (nums1[i] < nums2[j]) {
        i++;
      } else {
        j++;
      }
    }
    return res;
  }

  vector<int> intersection_builtin(vector<int>& nums1, vector<int>& nums2) {
    vector<int> res;
    set<int> set1(nums1.begin(), nums1.end()), set2(nums2.begin(), nums2.end());
    set_intersection(set1.begin(), set1.end(), set2.begin(), set2.end(), back_inserter(res));
    return res;
  }

  vector<int> intersection_map(vector<int>& nums1, vector<int>& nums2) {
    unordered_map<int, int> count;
    vector<int> res;
    for (size_t i = 0; i < nums1.size(); i++) {
      count[nums1[i]] = 1;
    }
    for (size_t j = 0; j < nums2.size(); j++) {
      if (count[nums2[j]] != 0) {
        res.push_back(nums2[j]);
        count[nums2[j]]--;
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0349_Intersection_of_Two_Arrays.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.*;

public class Question_0349_Intersection_of_Two_Arrays {
    public int[] intersection(int[] nums1, int[] nums2) {
        if (nums1 == null || nums2 == null) {
            return nums1 == null ? nums2 : nums1;
        }
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        List<Integer> list = new ArrayList<>();
        int i = 0, j = 0, m = nums1.length, n = nums2.length;
        while (i < m && j < n) {
            if (nums1[i] == nums2[j]) {
                list.add(nums1[i]);
                while (i < m - 1 && nums1[i] == nums1[i + 1]) {
                    i++;
                }
                while (j < n - 1 && nums2[j] == nums2[j + 1]) {
                    j++;
                }
                i++; j++;
            } else if (nums1[i] < nums2[j]) {
                i++;
            } else {
                j++;
            }
        }
        int[] res = new int[list.size()];
        for (i = 0; i < list.size(); i++) {
            res[i] = list.get(i);
        }
        return res;
    }

    public int[] intersection2(int[] nums1, int[] nums2) {
        if (nums1 == null || nums2 == null) {
            return nums1 == null ? nums2 : nums1;
        }
        Set<Integer> set1 = new HashSet<>(), set2 = new HashSet<>();
        for (int n : nums1) {
            set1.add(n);
        }
        for (int n : nums2) {
            set2.add(n);
        }
        set1.retainAll(set2);
        int[] res = new int[set1.size()];
        int i = 0;
        for (int n : set1) {
            res[i++] = n;
        }
        return res;
    }
}
